django/db/migrations/state.py

```diff
@@ -107,6 +110,155 @@ def remove_model(self, app_label, model_name):
             # the cache automatically (#24513)
             self.apps.clear_cache()
 
+    def rename_model(self, app_label, old_name, new_name):
```

Comment:
What implications does moving these methods onto the `ProjectState` have memory wise? Will Python need to clone the methods when cloing the project state?

Reply:
Is this really an issue we should keep an eye on?

`boundmethod` copying are pretty cheap and optimized for AFAIK and the number of `ProjectState` objects created during `makemigrations` and `migrate` is relatively low (maybe a few 100s of them in large projects). The memory footprint of 10 extra `boundmethod` instances should be order of magnitudes smaller than what lives in `ProjectState.__dict__` anyway?

Reply:
That's the thing. I don't (didn't) know what implications these boundmethods have. If they're not worth mentioning, I think it's fine. And yes, they're likely much smaller than `ProjectState.__dict__`.